### ISSUE TITLE
Eval bug: LoRA inference crashes GGML_ASSERT((int)sched->hash_set.size >= graph->n_nodes + graph->n_leafs) failed

### ISSUE THREAD

**Issue by arthurwu4work** [ggml-org/llama.cpp#16555]
### Name and Version

version: 6745 (a31cf36a)
built with cc (Ubuntu 11.4.0-1ubuntu1~22.04) 11.4.0 for x86_64-linux-gnu

### Operating systems

Linux

### GGML backends

CUDA

### Hardware

 NVIDIA H100

### Models

Meta-Llama-3.2-3B-Instruct.gguf and finetuned lora model

### Problem description & steps to reproduce

I've tested several recent versions of llama.cpp, including:
- b6745
- b6739
- b6715
- b6710
- b6708
- b6672

However, I consistently encounter the same error:
`llama.cpp/ggml/src/ggml-backend.cpp:1718: GGML_ASSERT((int)sched->hash_set.size >= graph->n_nodes + graph->n_leafs) failed`

The CLI command I'm running is:
`build/bin/llama-cli -m model/Meta-Llama-3.2-3B-Instruct.gguf --lora model/TEST_3B_1.1.20_192-NB_lora.gguf -sys "You are a help assistant"
`

If I run the base model without LoRA, everything works fine
`build/bin/llama-cli -m model/Meta-Llama-3.2-3B-Instruct.gguf -sys "You are a help assistant"`
But as soon as I include the LoRA model, the same error occurs.

Any idea what might be causing this or how to resolve it?

### Relevant log output

```shell
llama_adapter_lora_init_impl: loading lora adapter from 'model/TEST_3B_1.1.20_192-NB_lora.gguf' ...
llama_adapter_lora_init_impl: Dumping metadata keys/values.
llama_adapter_lora_init_impl: - kv   0:                       general.architecture str              = llama
llama_adapter_lora_init_impl: - kv   1:                               general.type str              = adapter
llama_adapter_lora_init_impl: - kv   2:                               adapter.type str              = lora
llama_adapter_lora_init_impl: - kv   3:                               general.name str              = Checkpoint 55
llama_adapter_lora_init_impl: - kv   4:                            general.version str              = 55
llama_adapter_lora_init_impl: - kv   5:                           general.basename str              = checkpoint
llama_adapter_lora_init_impl: - kv   6:                   general.base_model.count u32              = 1
llama_adapter_lora_init_impl: - kv   7:                  general.base_model.0.name str              = Meta Llama 3.2 3B Instruct
llama_adapter_lora_init_impl: - kv   8:                         adapter.lora.alpha f32              = 32.000000
llama_adapter_lora_init_impl: - kv   9:               general.quantization_version u32              = 2
llama_adapter_lora_init_impl:      CUDA0 LoRA buffer size =   185.50 MiB
llama_adapter_lora_init_impl: loaded 392 tensors from lora file
common_init_from_params: added <|end_of_text|> logit bias = -inf
common_init_from_params: added <|eom_id|> logit bias = -inf
common_init_from_params: added <|eot_id|> logit bias = -inf
common_init_from_params: setting dry_penalty_last_n to ctx_size = 4096
common_init_from_params: warming up the model with an empty run - please wait ... (--no-warmup to disable)
/root/llama.cpp_gpu_b6708/llama.cpp/ggml/src/ggml-backend.cpp:1718: GGML_ASSERT((int)sched->hash_set.size >= graph->n_nodes + graph->n_leafs) failed
/root/llama.cpp_gpu_b6708/llama.cpp/build/bin/libggml-base.so(+0x183cb)[0x7f153224f3cb]
/root/llama.cpp_gpu_b6708/llama.cpp/build/bin/libggml-base.so(ggml_print_backtrace+0x21f)[0x7f153224f82f]
/root/llama.cpp_gpu_b6708/llama.cpp/build/bin/libggml-base.so(ggml_abort+0x152)[0x7f153224fa02]
/root/llama.cpp_gpu_b6708/llama.cpp/build/bin/libggml-base.so(ggml_backend_sched_alloc_graph+0x1c3)[0x7f153226a113]
/root/llama.cpp_gpu_b6708/llama.cpp/build/bin/libllama.so(_ZN13llama_context14process_ubatchERK12llama_ubatch14llm_graph_typeP22llama_memory_context_iR11ggml_status+0xf3)[0x7f15323897f3]
/root/llama.cpp_gpu_b6708/llama.cpp/build/bin/libllama.so(_ZN13llama_context6decodeERK11llama_batch+0x2e7)[0x7f15323900b7]
/root/llama.cpp_gpu_b6708/llama.cpp/build/bin/libllama.so(llama_decode+0x10)[0x7f1532390f50]
build/bin/llama-cli(+0x158935)[0x558cbaa5b935]
build/bin/llama-cli(+0x442a3)[0x558cba9472a3]
/lib/x86_64-linux-gnu/libc.so.6(+0x29d90)[0x7f1531c55d90]
/lib/x86_64-linux-gnu/libc.so.6(__libc_start_main+0x80)[0x7f1531c55e40]
build/bin/llama-cli(+0x49c15)[0x558cba94cc15]
Aborted (core dumped)
```

**Comment by ggerganov** [comment]
Provide links to a model and a LoRA adapter.

**Comment by raja-gamevrs** [comment]
same issue was reported last week, #16475 . @ggerganov you can find them here
# Download base model and adapter
huggingface-cli download gameversellc/hermes_3_tests_gv Hermes-3-Llama-3.2-3B_q4_0.gguf
huggingface-cli download gameversellc/hermes_3_adapter_tests_gv gandalf_Hermes-3-Llama-3.2-3B_adapter.gguf

**Comment by arthurwu4work** [comment]
Hi @ggerganov ,

The base model and LoRA adapter files can be accessed via the following links.
[google drive](https://drive.google.com/file/d/1Z7FEJSYYWyuuW_b40Nx1FcCNo3Ymq_R8/view?usp=sharing)

Thank you.

**Comment by extra-large-onions** [comment]
any update on this? I encountered the same thing using unsloth's orpheus script https://github.com/unslothai/notebooks/blob/main/nb/Orpheus_(3B)-TTS.ipynb
when I run the test eval script within the notebook, it works fine 
<img width="682" height="389" alt="Image" src="https://github.com/user-attachments/assets/4f41a9c3-8e45-486f-94e8-9c8a50930722" />

but when I use the .gguf with convert_lora_to_gguf.py: 
ggml-backend.cpp:1634: GGML_ASSERT((int)sched->hash_set.size >= graph->n_nodes + graph->n_leafs) failed
my cmd: 
.\llama-server.exe -m ./orpheus-3b-0.1-ft-Q4_K_M.gguf --lora ./Lora_Dragonfox_60-F16-LoRA.gguf --ctx-size 4096 --port 8020
llama.cpp version - tried multiple (cpu only, gpu) versions (b6229, b6397, b6922)
it works fine when i remove the lora args

**Comment by mimi3421** [comment]
It seems that the calculation of `sched->hash_set.size` has been changed somewhere after tag ```b6000``` from 65537 to a small value 2000 or so. 

For testing, I have made a debug release of ```llama.cpp``` showing this change ( https://github.com/mimi3421/llama.cpp_test/releases/tag/v0.1 ).

``` bash
set GGML_DEBUG=1
llama-server --jinja -fa on --pooling last --model "phi-4-IQ3_M.gguf" --lora "LoRA-phi-4-abliterated-f16.gguf" -c 3072 --threads 6 --n-gpu-layers 999

# Newer version with Lora (Crashed):
ggml_backend_sched_alloc_graph: hash_set.size: 2053, n_nodes: 1815, n_leafs: 653
D:/a/llama.cpp_test/llama.cpp_test/ggml/src/ggml-backend.cpp:1719: GGML_ASSERT((int)sched->hash_set.size >= graph->n_nodes + graph->n_leafs) failed

# Newer version without Lora:
ggml_backend_sched_alloc_graph: hash_set.size: 2053, n_nodes: 1167, n_leafs: 329
...

llama-server --jinja -fa --pooling last --model "phi-4-IQ3_M.gguf" --lora "LoRA-phi-4-abliterated-f16.gguf" -c 3072 --threads 6 --n-gpu-layers 999

# Older version with Lora:
ggml_backend_sched_alloc_graph: hash_set.size: 65537, n_nodes: 2175, n_leafs: 651
...

# Older version without Lora:
ggml_backend_sched_alloc_graph: hash_set.size: 65537, n_nodes: 1527, n_leafs: 327
...
```

**Comment by ggerganov** [comment]
Could you track the exact commit that breaks this?

**Comment by mimi3421** [comment]
@ggerganov 

Sure. My device is 64GB ram with RTX2070 (8GB VRAM). The model with LoRA consumes approximately 7.6GB VRAM without using the shared memory in its normal run. I run the following script with ```llama-*-bin-win-vulkan-x64.zip```:

``` bash
llama-server --jinja -fa --pooling last --model "phi-4-IQ3_M.gguf" --lora "LoRA-phi-4-abliterated-f16.gguf" -c 3072 --threads 6 --n-gpu-layers 999
```

And find that 

1. The last build with normal function is ```b5932```.
2. Start from ```b5933``` to ```b6070```, the program failed with the following message:

    > llama_adapter_lora_init_impl: loading lora adapter from 'LoRA-phi-4-abliterated-f16.gguf' ...
    > llama_adapter_lora_init_impl:    Vulkan0 LoRA buffer size =   437.88 MiB
    > llama_adapter_lora_init_impl: CPU_Mapped LoRA buffer size =    12.88 MiB
    > llama_adapter_lora_init_impl: loaded 324 tensors from lora file
    > common_init_from_params: added <|endoftext|> logit bias = -inf
    > common_init_from_params: added <|im_end|> logit bias = -inf
    > common_init_from_params: setting dry_penalty_last_n to ctx_size = 3072
    > common_init_from_params: warming up the model with an empty run - please wait ... (--no-warmup to disable)
    > ggml_new_object: not enough space in the context's memory pool (needed 796704, available 796336)
    > C:/a/llama.cpp/llama.cpp/ggml/src/ggml.c:1649: GGML_ASSERT(obj_new) failed

3.  Start from ```b6071``` to the most recent version, the program failed with the following message:

    > llama_adapter_lora_init_impl: loading lora adapter from 'LoRA-phi-4-abliterated-f16.gguf' ...
    > llama_adapter_lora_init_impl:    Vulkan0 LoRA buffer size =   437.88 MiB
    > llama_adapter_lora_init_impl: CPU_Mapped LoRA buffer size =    12.88 MiB
    > llama_adapter_lora_init_impl: loaded 324 tensors from lora file
    > common_init_from_params: added <|endoftext|> logit bias = -inf
    > common_init_from_params: added <|im_end|> logit bias = -inf
    > common_init_from_params: setting dry_penalty_last_n to ctx_size = 3072
    > common_init_from_params: warming up the model with an empty run - please wait ... (--no-warmup to disable)
    > C:/a/llama.cpp/llama.cpp/ggml/src/ggml-backend.cpp:1551: GGML_ASSERT((int)sched->hash_set.size >= graph->n_nodes + graph->n_leafs) failed

Hope these information help.

**Comment by mimi3421** [comment]
It seems that the bug locates at the definition of ```llama_context::graph_max_nodes``` in ```src/llama-context.cpp```. ( https://github.com/ggml-org/llama.cpp/compare/b5932...b5933 ). Changing the maximum graph nodes back to 65536 or some larger one (larger than graph->n_nodes + graph->n_leafs +1 ? ) would solve the problem.

The ```model.n_tensors()``` in my case is 243, while with Lora, the ```graph->n_nodes + graph->n_leafs``` is 2826, which is approaching ```12u*model.n_tensors()```. Maybe ```std::max<uint32_t>(1024u, 8u*model.n_tensors())``` is not a good way to approximate the needed number of graph nodes .

``` C
uint32_t llama_context::graph_max_nodes() const {
    - return std::max<uint32_t>(1024u, 8u*model.n_tensors());
    + return std::max<uint32_t>(65536u, 5u*model.n_tensors());
}
```

**Comment by extra-large-onions** [comment]
>     1. The last build with normal function is `b5932`.

can confirm that 5932 works, thanks a lot for the workaround

**Comment by ggerganov** [comment]
> Maybe std::max<uint32_t>(1024u, 8u*model.n_tensors()) is not a good way to approximate the needed number of graph nodes .

There isn't a good way to determine this in advance. I guess we can bump the factor to `16u` for now.

**Comment by nendraharyo** [comment]
> It seems that the bug locates at the definition of `llama_context::graph_max_nodes` in `src/llama-context.cpp`. ( [b5932...b5933](https://github.com/ggml-org/llama.cpp/compare/b5932...b5933) ). Changing the maximum graph nodes back to 65536 or some larger one (larger than graph->n_nodes + graph->n_leafs +1 ? ) would solve the problem.
> 
> The `model.n_tensors()` in my case is 243, while with Lora, the `graph->n_nodes + graph->n_leafs` is 2826, which is approaching `12u*model.n_tensors()`. Maybe `std::max<uint32_t>(1024u, 8u*model.n_tensors())` is not a good way to approximate the needed number of graph nodes .
> 
> uint32_t llama_context::graph_max_nodes() const {
>     - return std::max<uint32_t>(1024u, 8u*model.n_tensors());
>     + return std::max<uint32_t>(65536u, 5u*model.n_tensors());
> }

having the same problem, changed this and build locally fixed it
